src/main/java/net/cardosi/mojo/cache/FileService.java

```diff
@@ -0,0 +1,136 @@
+
+    public FileService(DiskCache diskCache, CachedProject... cachedProjects) {
+        int                            watchSize       = 0;
+        Map<CachedProject, List<Path>> projectsToWatch = new HashMap<>();
+        this.diskCache = diskCache;
+
+
+        List<CachedProject> cachedProjectsList = new ArrayList<>();
+
+        // need all reactor projects
+        for (CachedProject project : cachedProjects) {
+            cachedProjectsList.add(project);
+            cachedProjectsList.addAll(project.getChildren().stream().filter( p -> p.hasSourcesMapped() ).collect(Collectors.toList()));
+        }
+
+        for (CachedProject project : cachedProjectsList) {
+            project.compileSourceRoots().stream().map(Paths::get).forEach(p -> projects.put(p, project));
+        }
+
+        this.listener = new ChangeSetListener();
+        try {
+            this.watcher = DirectoryWatcher.builder()
+                                           .paths(new ArrayList<>(projects.keySet()))
+                                           .listener(listener)
+                                           .fileHashing(true)
+                                           .build();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void start() {
+        watcher.watchAsync();
+
+        UserInputRunner inputRunner = new UserInputRunner(watcher, diskCache, run,
+                                                          listener, projects);
+        new Thread(inputRunner).start();
+    }
+
+    public void stop() {
+        // guarantee this goes false, so all threads stop.
+        while ( run.compareAndSet(true,false) ) {
+            try {
+                Thread.sleep(50);
+            } catch (InterruptedException e) {
+            }
+        }
+
+        try {
+            this.watcher.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static class UserInputRunner implements Runnable {
+        private final AtomicBoolean            run;
+        private final ChangeSetListener        listener;
+        private final Map<Path, CachedProject> projects;
+        private final DirectoryWatcher         watcher;
+        private final DiskCache                diskCache;
+
+        public UserInputRunner(DirectoryWatcher  watcher, DiskCache diskCache, AtomicBoolean run, ChangeSetListener listener, Map<Path, CachedProject> projects) {
+            this.watcher = watcher;
+            this.diskCache = diskCache;
+            this.run = run;
+            this.listener = listener;
+            this.projects = projects;
+        }
+
+        @Override
+        public void run() {
+            while (run.get()) {
+                try {
+                    System.out.println("waiting read");
+                    System.in.read();
+                    Map<Path, ChangeSet> changeSets = listener.getChangeSet();
```

Comment:
Could use a comment explaining the reason and usage.

Reply:
Wait why is a comment needed here? "user hit enter, get the changes since we last looked" then below those changes are iterated, marked dirty as appropriate?

https://github.com/gmethvin/directory-watcher/blob/master/core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java

Reply:
I added some more comments, not sorer if that's what you wanted.